src/rhsmlib/services/register.py

```diff
@@ -0,0 +1,100 @@
+from __future__ import print_function, division, absolute_import
+
+# Copyright (c) 2017 Red Hat, Inc.
+#
+# This software is licensed to you under the GNU General Public License,
+# version 2 (GPLv2). There is NO WARRANTY for this software, express or
+# implied, including the implied warranties of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+# along with this software; if not, see
+# http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+#
+# Red Hat trademarks are not licensed under GPLv2. No permission is
+# granted to use or replicate Red Hat trademarks that are incorporated
+# in this software or its documentation.
+import logging
+import socket
+
+from rhsmlib.services import exceptions
+
+from subscription_manager import injection as inj
+from subscription_manager.i18n import ugettext as _
+
+log = logging.getLogger(__name__)
+
+
+class RegisterService(object):
+    def __init__(self, cp):
+        self.plugin_manager = inj.require(inj.PLUGIN_MANAGER)
+        self.installed_mgr = inj.require(inj.INSTALLED_PRODUCTS_MANAGER)
+        self.facts = inj.require(inj.FACTS)
+        self.identity = inj.require(inj.IDENTITY)
+        self.cp = cp
+
+    def register(self, org, activation_keys=None, environment=None, force=None, name=None, consumerid=None,
+            **kwargs):
+        if kwargs:
+            raise exceptions.ValidationError(_("Unknown arguments: %s") % kwargs.keys())
+
+        options = {
+            'activation_keys': activation_keys,
+            'environment': environment,
+            'force': force,
+            'name': name,
+            'consumerid': consumerid
+        }
+        self.validate_options(options)
+
+        environment = options['environment']
+        facts_dict = self.facts.get_facts()
+
+        # Default to the hostname if no name is given
+        consumer_name = options['name'] or socket.gethostname()
+
+        self.plugin_manager.run("pre_register_consumer", name=consumer_name, facts=facts_dict)
+
+        if consumerid:
+            consumer = self.cp.getConsumer(consumerid)
+            if consumer.get('type', {}).get('manifest', {}):
+                raise exceptions.ServiceError(
+                    "Registration attempted with a consumer ID that is not of type 'system'")
+        else:
+            consumer = self.cp.registerConsumer(
+                name=consumer_name,
+                facts=facts_dict,
+                owner=org,
+                environment=environment,
+                keys=options.get('activation_keys'),
+                installed_products=self.installed_mgr.format_for_server(),
+                content_tags=self.installed_mgr.tags
+            )
```

Comment:
The identity cert (`self.identity` in this context) needs a call to `reload` at some point. Right after this statement or in `managerlib.persist_consumer_cert` might be appropriate. Without the reload, if I call register twice it doesn't acknowledge that I'm already registered, and lets me repeatedly call register.